# Synent Technologies – Task 1: Data Cleaning & Preprocessing

**Internship Domain:** Data Science  
**Task:** Data Cleaning & Preprocessing  
**Dataset:** Titanic Dataset

## Objective
Clean and prepare the Titanic dataset for analysis by:
- Handling missing values
- Removing duplicate rows
- Converting data types
- Renaming columns

The final output is a clean dataset ready for analysis.

## 1. Import Libraries

In [2]:
import pandas as pd
pd.set_option("display.max_columns",None)

## 2. Load the Dataset

In [3]:
df= pd.read_csv("Titanic-Dataset.csv")
print("Dataset shape:",df.shape)
df.head()

Dataset shape: (891, 12)


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


## 3. Inspect the Dataset

In [4]:
print("Column names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

print("\nDuplicate rows:",df.duplicated().sum())

Column names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

Data types:
PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

Missing values:
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Duplicate rows: 0


## 4. Handle Missing Values

The missing values are handled as follows:

- **Age:** missing values are replaced with the median age because it is less affected by extreme values.
- **Embarked:** missing values are replaced with the most frequent value (mode).
- **Cabin:** missing values are replaced with `"Unknown"` because many cabin values are unavailable, so dropping the column would lose useful information.

In [6]:
print("Missing values before cleaning:")
print(df.isnull().sum())
df["Age"]= df["Age"].fillna(df["Age"].median())
df["Embarked"] =df["Embarked"].fillna(df["Embarked"].mode()[0])
df["Cabin"]= df["Cabin"].fillna("Unknown")

print("\nMissing values after handling:")
print(df.isnull().sum())

Missing values before cleaning:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

Missing values after handling:
PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64


## 5. Remove Duplicate Rows

In [8]:
duplicates_before= df.duplicated().sum()
print("Duplicate rows before removal:",duplicates_before)
df= df.drop_duplicates().copy()
print("Duplicate rows after removal:",df.duplicated().sum())

Duplicate rows before removal: 0
Duplicate rows after removal: 0


## 6. Convert Data Types

Numeric columns are explicitly converted to appropriate numeric types. This makes the dataset consistent and ready for later analysis.

In [9]:
df["PassengerId"]= df["PassengerId"].astype("int64")
df["Survived"]= df["Survived"].astype("int64")
df["Pclass"]= df["Pclass"].astype("int64")
df["SibSp"]= df["SibSp"].astype("int64")
df["Parch"]= df["Parch"].astype("int64")
df["Fare"]= df["Fare"].astype("float64")
print(df.dtypes)

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object


## 7. Rename Columns

In [11]:
df = df.rename(columns={
    "PassengerId":"passenger_id",
    "Survived":"survived",
    "Pclass":"passenger_class",
    "Name":"name",
    "Sex":"sex",
    "Age":"age",
    "SibSp":"siblings_spouses",
    "Parch":"parents_children",
    "Ticket":"ticket",
    "Fare":"fare",
    "Cabin":"cabin",
    "Embarked":"embarked"
})

print("New column names:")
print(df.columns.tolist())

New column names:
['passenger_id', 'survived', 'passenger_class', 'name', 'sex', 'age', 'siblings_spouses', 'parents_children', 'ticket', 'fare', 'cabin', 'embarked']


## 8. Verify the Cleaned Dataset

In [12]:
print("Final dataset shape:",df.shape)
print("Total missing values:",df.isnull().sum().sum())
print("Total duplicate rows:",df.duplicated().sum())
df.head()

Final dataset shape: (891, 12)
Total missing values: 0
Total duplicate rows: 0


,passenger_id,survived,passenger_class,name,sex,age,siblings_spouses,parents_children,ticket,fare,cabin,embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,Unknown,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,Unknown,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,Unknown,S


## 9. Save the Cleaned Dataset

The cleaned dataset is saved as `titanic_cleaned.csv` and is ready for further analysis.

In [13]:
df.to_csv("titanic_cleaned.csv",index=False)
print("Cleaned dataset saved successfully as titanic_cleaned.csv")

Cleaned dataset saved successfully as titanic_cleaned.csv


## 10. Final Summary

After preprocessing:

- All missing values were handled.
- Duplicate rows were removed.
- Data types were reviewed and converted where needed.
- Column names were renamed using a consistent format.
- The final dataset contains **891 rows and 12 columns**.
- The final dataset contains **0 missing values** and **0 duplicate rows**.

The cleaned dataset is ready for analysis.